Convert top-level dashboard components to TypeScript

kulyk opened the pull request:
Converts top-level dashboard components to TypeScript. That should make other dashboard components a bit more reliable as TypeScript would know which props and redux state exist in top-level dashboard components.

Would be easier to review commit-by-commit:

- [convert `DashboardApp` to TypeScript](https://github.com/metabase/metabase/commit/ea30674e9fc43ddd113f809fae707f903a92248c)
- [convert `Dashboard` to TypeScript](https://github.com/metabase/metabase/commit/a4449f6d867910b3c801645d27746cd5e009c5d3)
- [convert `DashboardGrid` to TypeScript](https://github.com/metabase/metabase/commit/3ce5cc3a323add349cfd25b5a71a470976cea773)
- [convert `DashboardHeader` to TypeScript](https://github.com/metabase/metabase/commit/c4254d42fe123851acc3f9710cd7ae637333c906)

**To verify:** CI should be green 🟢

kulyk committed:
Convert `DashboardApp` to TypeScript

kulyk committed:
Convert `Dashboard` to TypeScript

kulyk committed:
Convert `DashboardGrid` to TypeScript

kulyk committed:
Convert `DashboardHeader` to TypeScript

kulyk committed:
Bind `setDashboardAttribute` method

kulyk committed:
Fix adding dashcards

uladzimirdev commented:
<details><summary>Holy cow!</summary>
<p>

![image](https://media.giphy.com/media/hWAEptmZvfQkrJHTOk/giphy.gif)

</p>
</details>

kulyk committed:
Fix question's "Add to dashboard" flow

oleggromov commented on frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx:
```diff
@@ -25,12 +59,127 @@ import {
+    id: DashCardId,
+    columnKey: string,
+    settings?: Record<string, unknown> | null,
+  ) => void;
```
👏

oleggromov commented on frontend/src/metabase/dashboard/components/DashboardHeader/DashboardHeader.tsx:
```diff
@@ -45,16 +46,129 @@ import { dismissAllUndo } from "metabase/redux/undo";
+  onNightModeChange: () => void;
+
+  setSidebar: (opts: { name: DashboardSidebarName }) => void;
+  closeSidebar: () => void;
```
👏

kamilmielnik commented on frontend/src/metabase/dashboard/components/DashCard/DashCard.tsx:
```diff
@@ -79,6 +79,15 @@ export interface DashCardProps {
   navigateToNewCardFromDashboard?: (
     opts: NavigateToNewCardFromDashboardOpts,
   ) => void;
+  fetchCardData: (
```
This prop is never used, we should not be passing it in `DashboardGrid` or unit tests.

kamilmielnik commented on frontend/src/metabase/dashboard/components/DashboardHeader/DashboardHeaderView.tsx:
```diff
@@ -27,10 +27,10 @@ import { DashboardTabs } from "../../components/DashboardTabs/DashboardTabs";
 
 interface DashboardHeaderViewProps {
   editingTitle: string;
```
There's a default value for it

```suggestion
  editingTitle?: string;
```

kamilmielnik commented on frontend/src/metabase/dashboard/components/DashboardHeader/DashboardHeader.tsx:
```diff
@@ -45,16 +46,129 @@ import { dismissAllUndo } from "metabase/redux/undo";
+      preserveParameters?: boolean;
+    };
+  }) => Promise<void>;
+  fetchPulseFormInput: () => void;
```
This should have been defined in `DispatchProps`, not `OwnProps`

kamilmielnik commented on frontend/src/metabase/dashboard/components/DashboardGrid.tsx:
```diff
@@ -41,12 +56,104 @@ import { AddSeriesModal } from "./AddSeriesModal/AddSeriesModal";
+
+  onChangeLocation: (location: LocationDescriptor) => void;
+
+  // TODO
+  showClickBehaviorSidebar: () => void;
+  navigateToNewCardFromDashboard: () => void;
+  markNewCardSeen: () => void;
```
Leftover TODO

kamilmielnik commented on frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx:
```diff
@@ -163,7 +315,7 @@ class DashboardInner extends Component {
   };
 
-  onMainScroll = event => {
+  onMainScroll = (event: any) => {
```
```suggestion
  onMainScroll = (event: Event) => {
```

kulyk replied:
Not sure how we should type this 🤔 `Event` doesn't seem to be aware of the `target` element or about the fact that we're scrolling, so it doesn't have the `event.target.scrollTop` property that we need. Couldn't find a proper type for this and tried to workaround it with an explicit type check, but still no luck:

```ts
if (
  event.target &&
  'scrollTop' in event.target &&
  typeof event.target.scrollTop === "number"
) {
  event.target.scrollTop // still doesn't work 😢
}
```

WDYT?

kamilmielnik replied:
`getMainElement()` returns `HTMLElement` so we can do this:

```tsx
  onMainScroll = (event: Event) => {
    if (event.target instanceof HTMLElement) {
      this.setState({ hasScroll: event.target.scrollTop > 0 });
    }
  };
```

or even:
```tsx
  onMainScroll = (event: Event) => {
    if (!(event.target instanceof HTMLElement)) {
      // sanity check, this should never happen
      throw new Error("main element is not an HTMLElement");
    }

    this.setState({ hasScroll: event.target.scrollTop > 0 });
  };
```

kamilmielnik commented on frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx:
```diff
@@ -124,10 +272,11 @@ class DashboardInner extends Component {
       }
-    } catch (error) {
+    } catch (err) {
+      const error = err as any;
       if (error.status === 404) {
```
```suggestion
      if (error instance of Response && error.status === 404) {
```

or please add a custom type guard

kamilmielnik commented on frontend/src/metabase/dashboard/components/DashboardGrid.tsx:
```diff
@@ -41,12 +56,104 @@ import { AddSeriesModal } from "./AddSeriesModal/AddSeriesModal";
+interface DashboardGridState {
+  visibleCardIds: Set<number>;
+  initialCardSizes: { [key: string]: { w: number; h: number } };
+  layouts: { desktop: any[]; mobile: any[] };
+  addSeriesModalDashCard: any;
+  replaceCardModalDashCard: any;
```
Leftover `any`s

kamilmielnik commented on frontend/src/metabase/dashboard/components/DashboardGrid.tsx:
```diff
@@ -41,12 +56,104 @@ import { AddSeriesModal } from "./AddSeriesModal/AddSeriesModal";
-  constructor(props, context) {
+  _pauseAnimationTimer: ReturnType<typeof setTimeout> | null = null;
+
+  constructor(props: DashboardGridProps, context: any) {
```
```suggestion
  constructor(props: DashboardGridProps, context: unknown) {
```

kulyk committed:
Remove `fetchCardData` prop from `DashCard` (not used)

kulyk committed:
Make `editingTitle` prop optional

kulyk committed:
Move `fetchPulseFormInput` to `DispatchProps`

kulyk committed:
Fix TODO types in `DashboardGrid`

kulyk committed:
Fix error handling

kulyk committed:
Replace `any` with `unknown`

kulyk committed:
Fix `any` in `DashboardGrid`

kulyk commented:
Thanks for such a thorough review @kamilmielnik 😍 I think I've addressed all of your comments except for [that one](https://github.com/metabase/metabase/pull/37703#discussion_r1457399372), could you please take a look and maybe you happen to have a good solution?

kamilmielnik reviewed (approved):
This is awesome! :heart:

kulyk committed:
Fix `onMainScroll` type
